uat/mqtt-client-control/src/main/java/com/aws/greengrass/testing/mqtt/client/control/implementation/AgentControlImpl.java

```diff
@@ -0,0 +1,248 @@
+                throw new RuntimeException("Couldn't create MQTT connection: " + error);
+            }
+        } finally {
+            connectLock.unlock();
+        }
+        logger.atInfo().log("createMqttConnection: MQTT connectionId {} created", connectionControl.getConnectionId());
+        return connectionControl;
+    }
+
+    /**
+     * Close MQTT connection to the broker.
+     *
+     * @param closeRequest parameters of MQTT disconnect
+     * @throws StatusRuntimeException on errors
+     */
+    public void closeMqttConnection(MqttCloseRequest closeRequest) {
+        blockingStub.closeMqttConnection(closeRequest);
+        int connectionId = closeRequest.getConnectionId().getConnectionId();
+        connections.remove(connectionId);
+        logger.atInfo().log("closeMqttConnection: MQTT connectionId {} closed", connectionId);
+    }
+
+    /**
+     * Do MQTT subscription(s).
+     *
+     * @param subscribeRequest subscribe request
+     * @return reply to subscribe
+     * @throws StatusRuntimeException on errors
+     */
+    public MqttSubscribeReply subscribeMqtt(MqttSubscribeRequest subscribeRequest) {
+        int connectionId = subscribeRequest.getConnectionId().getConnectionId();
+        logger.atInfo().log("SubscribeMqtt: subscribe on connection {}", connectionId);
+        return blockingStub.subscribeMqtt(subscribeRequest);
+    }
+
+    /**
+     * Remove MQTT subscription(s).
+     *
+     * @param unsubscribeRequest unsubscribe request
+     * @return reply to unsubscribe
+     * @throws StatusRuntimeException on errors
+     */
+    public MqttSubscribeReply unsubscribeMqtt(MqttUnsubscribeRequest unsubscribeRequest) {
+        int connectionId = unsubscribeRequest.getConnectionId().getConnectionId();
+        logger.atInfo().log("UnsubscribeMqtt: unsubscribe on connectionId {}", connectionId);
+        return blockingStub.unsubscribeMqtt(unsubscribeRequest);
+    }
+
+    /**
+     * Publish MQTT message.
+     *
+     * @param publishRequest publish request
+     * @return reply to publish
+     * @throws StatusRuntimeException on errors
+     */
+    public MqttPublishReply publishMqtt(MqttPublishRequest publishRequest) {
+        int connectionId = publishRequest.getConnectionId().getConnectionId();
+        String topic = publishRequest.getMsg().getTopic();
+        logger.atInfo().log("PublishMqtt: publishing on connectionId {} topic {}", connectionId, topic);
+        return blockingStub.publishMqtt(publishRequest);
+    }
+
+    private void closeAllMqttConnections() {
+        connections.forEach((connectionId, connectionControl) -> {
+            if (connections.remove(connectionId, connectionControl)) {
+                connectionControl.closeMqttConnection(DEFAULT_DISCONNECT_REASON);
+            }
+        });
+    }
+
+    private void disconnect() {
+        if (! isDisconnected.getAndSet(true)) {
+            if (! isShutdownSent.getAndSet(true)) {
```

Comment:
no spaces after !

Reply:
Fixed in 88b9536b76b151049a1931240b9578cd338825ec